packages/pds/src/db/message-queue/index.ts

```diff
@@ -0,0 +1,288 @@
+  AddUpvote,
+  CreateNotification,
+  DeleteNotifications,
+  Message,
+  RemoveMember,
+  RemoveUpvote,
+} from './messages'
+import { sql } from 'kysely'
+import { dbLogger as log } from '../../logger'
+import { AuthStore } from '@atproto/auth'
+import * as repoUtil from '../../util/repo'
+import * as schema from '../../lexicon/schemas'
+import { TID } from '@atproto/common'
+import { MessageQueue } from '../types'
+
+type GetAuthStoreFn = (did: string) => AuthStore
+
+export class SqlMessageQueue implements MessageQueue {
+  private cursorExists = false
+
+  constructor(
+    private name: string,
+    private db: Database,
+    private getAuthStore: GetAuthStoreFn,
+  ) {}
+
+  async send(tx: Database, messages: Message | Message[]): Promise<void> {
+    const msgArray = Array.isArray(messages) ? messages : [messages]
+    if (msgArray.length === 0) return
+    const now = new Date().toISOString()
+    const values = msgArray.map((msg) => ({
+      message: JSON.stringify(msg),
+      createdAt: now,
+    }))
+
+    await tx.db.insertInto('message_queue').values(values).execute()
+
+    this.process()
+  }
+
+  private async ensureCursor(): Promise<void> {
+    if (this.cursorExists) return
+    await this.db.db
+      .insertInto('message_queue_cursor')
+      .values({ consumer: this.name, cursor: 1 })
+      .onConflict((oc) => oc.doNothing())
+      .execute()
+    this.cursorExists = true
+  }
+
+  async process(): Promise<void> {
+    await this.ensureCursor()
+    const keepGoing = await this.db.transaction(
+      async (dbTxn): Promise<boolean> => {
+        let builder = dbTxn.db
+          .selectFrom('message_queue_cursor as cursor')
+          .innerJoin('message_queue', 'message_queue.id', 'cursor.cursor')
+          .selectAll()
+        if (this.db.dialect !== 'sqlite') {
+          builder = builder.forUpdate()
+        }
+
+        const res = await builder.executeTakeFirst()
+        // all caught up
+        if (!res) return false
+
+        const message: Message = JSON.parse(res.message)
+        await this.handleMessage(dbTxn, message)
+        await dbTxn.db
+          .updateTable('message_queue_cursor')
+          .set({ cursor: sql`cursor + 1` })
+          .where('consumer', '=', this.name)
+          .execute()
+        return true
+      },
+    )
+    if (keepGoing) {
+      await this.process()
+    }
+  }
+
+  private async handleMessage(db: Database, message: Message) {
+    switch (message.type) {
+      case 'add_member':
+        return this.handleAddMember(db, message)
+      case 'remove_member':
+        return this.handleRemoveMember(db, message)
+      case 'add_upvote':
+        return this.handleAddUpvote(db, message)
+      case 'remove_upvote':
+        return this.handleRemoveUpvote(db, message)
+      case 'create_notification':
+        return this.handleCreateNotification(db, message)
+      case 'delete_notifications':
+        return this.handleDeleteNotifications(db, message)
+    }
+  }
```

Comment:
Eventually I think it might be nice to colocate these handlers with related code (e.g.`handleAddMember()` living closer to member methods) and allow other areas of the app to hook-in to the message queue.  Thinking of the message queue as more of a pull-based transport, rather than a vertical slice of the app.  Zero problem with it as-is today, though, just thinkin!

Reply:
yup makes sense 👍